### PR TITLE
Infer dtype in SymbolBlock import from input symbol

### PR DESCRIPTION
**Pull request by sandeep-krishnamurthy**
## Description ##
* Fix for the issue - https://github.com/apache/incubator-mxnet/issues/11849 
* Currently, Gluon symbol block cannot import any symbol with type other than fp32. All the parameters are created as FP32 leading to failure in importing the params when it is of type fp16, fp64 etc,
* In this PR, we infer the type of the symbol being imported and create the Symbol Block Parameters with that inferred type.

~~Created this PR for getting the early feedback. I am working on adding test cases for this and will update the PR soon~~ 
Added the tests

@szha @hetong007 @apeforest - Can you please take a look at this?
@Roshrini - I think this is important fix to be picked for 1.3. @szha ?

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [X] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
* In this PR, we infer the type of the symbol being imported and create the Symbol Block Parameters with that inferred type.
* If we cannot infer the type, we go by usual route using default dtype (fp32) for creating the parameters.

### Comments ###
Below is the issue:

When you create mx.gluon.SymbolBlock(sm, input). It creates the parameters in the Block. Type is not passed for creating the parameter.
See here - https://github.com/apache/incubator-mxnet/blob/master/python/mxnet/gluon/block.py#L1058 and the behavior is "If there is not parameter to get, it creates one and uses default type (fp32) https://github.com/apache/incubator-mxnet/blob/master/python/mxnet/gluon/parameter.py#L688

### CODE REVIEW

File: python/mxnet/gluon/block.py [hunk 1]
```diff
@@ -1053,13 +1053,34 @@ def __init__(self, outputs, inputs, params=None):
...
+            for i, arg in enumerate(arg_params):
+                if arg not in input_names:
+                    self.params.get(arg, allow_deferred_init=True, dtype=arg_types[i])
+
+            for i, aux in enumerate(aux_params):
+                if aux not in input_names:
+                    self.params.get(aux, grad_req='null', allow_deferred_init=True, dtype=aux_types[i])
+        else:
+            # Use default types for params
+            for i, arg in enumerate(arg_params):
```
Comment (apeforest) [inline python/mxnet/gluon/block.py]: Maybe no enumerate is needed since you don't use index i in this loop?

File: python/mxnet/gluon/block.py [hunk 2]
```diff
@@ -1053,13 +1053,34 @@ def __init__(self, outputs, inputs, params=None):
...
+                if arg not in input_names:
+                    self.params.get(arg, allow_deferred_init=True)
 
-        for i in out.list_auxiliary_states():
-            if i not in input_names:
-                self.params.get(i, grad_req='null', allow_deferred_init=True)
+            for i, aux in enumerate(aux_params):
```
Comment (apeforest) [inline python/mxnet/gluon/block.py]: Same here, no need to get index i

File: python/mxnet/gluon/block.py [hunk 3]
```diff
@@ -1086,3 +1107,49 @@ def _clear_cached_op(self):
...
+
+    # Try to infer types of other parameters.
+    if input_sym_arg_type and len(input_sym_arg_type) > 0:
+        params = {input_sym_name:input_sym_arg_type[0]}
+        arg_types, _, aux_types = out_params.infer_type(**params)
+        if arg_types is not None and len(arg_types) == len(arg_params) and \
+           aux_types is not None and len(aux_types) == len(aux_params):
+            infer_type_success = True
+
+    return (infer_type_success, arg_types, aux_types)
```
Comment (szha) [inline python/mxnet/gluon/block.py]: if arg_types or aux_types are None, just fill it with default type. feel free to add a default_type argument to this function to make it flexible.

File: tests/python/unittest/test_gluon.py [hunk 4]
```diff
@@ -336,6 +339,28 @@ def hybrid_forward(self, F, x):
...
 
+    # Test case to verify if initializing the SymbolBlock from a model with params 
+    # other than fp32 param dtype.
+    # Load a resnet model, cast it to fp64 and export
+    tmp = tempfile.mkdtemp()
+    tmpfile = os.path.join(tmp, 'resnet34_fp64')
+    ctx = mx.cpu(0)
+
+    net_fp32 = mx.gluon.model_zoo.vision.resnet34_v2(pretrained=True, ctx=ctx)
+    net_fp32.cast('float64')
```
Comment (lupesko) [inline tests/python/unittest/test_gluon.py]: Shall we add another test that casts to float16?
The original issue reported import from float16 failing, and it might be appropriate to cover it as well.

File: python/mxnet/gluon/block.py (hunk 3 above)
Comment (sandeep-krishnamurthy) [reply python/mxnet/gluon/block.py]: Done.

File: tests/python/unittest/test_gluon.py (hunk 4 above)
Comment (sandeep-krishnamurthy) [reply tests/python/unittest/test_gluon.py]: Added a test for fp16.

File: python/mxnet/gluon/block.py (hunk 2 above)
Comment (sandeep-krishnamurthy) [reply python/mxnet/gluon/block.py]: N/A with latest refactoring.

**Comment by sandeep-krishnamurthy** [comment]
@apeforest @szha @lupesko - Thanks for your time. Addressed your comments. Please have a look.

File: tests/python/unittest/test_gluon.py [hunk 5]
```diff
@@ -336,6 +339,29 @@ def hybrid_forward(self, F, x):
...
 
+    # Test case to verify if initializing the SymbolBlock from a model with params 
+    # other than fp32 param dtype.
+
+    # 1. Load a resnet model, cast it to fp64 and export
+    tmp = tempfile.mkdtemp()
+    tmpfile = os.path.join(tmp, 'resnet34_fp64')
+    ctx = mx.cpu(0)
+
+    net_fp32 = mx.gluon.model_zoo.vision.resnet34_v2(pretrained=True, ctx=ctx, root=tmp)
```
Comment (apeforest) [inline tests/python/unittest/test_gluon.py]: Is the model name in model zoo going to be maintained? If there is any change in the name, it would break this unit test. Not sure if we want to keep this dependency

Comment (sandeep-krishnamurthy) [reply tests/python/unittest/test_gluon.py]: resnet34_v2 is a public function exposed through model_zoo.vision module. I think it is ok because we are not using string based selection of the model.

**Comment by sandeep-krishnamurthy** [comment]
@szha @apeforest - Is this good to go?

File: tests/python/unittest/test_gluon.py [hunk 6]
```diff
@@ -336,6 +339,34 @@ def hybrid_forward(self, F, x):
     net.hybridize()
     assert isinstance(net(mx.nd.zeros((16, 10))), mx.nd.NDArray)
 
+    # Test case to verify if initializing the SymbolBlock from a model with params 
+    # other than fp32 param dtype.
+
+    # 1. Load a resnet model, cast it to fp64 and export
+    tmp = tempfile.mkdtemp()
```
Comment (stu1130) [inline tests/python/unittest/test_gluon.py]: should we delete the temporary directory when done with it?

Comment (sandeep-krishnamurthy) [reply tests/python/unittest/test_gluon.py]: Temp gets automatically cleaned up.

Comment (stu1130) [reply tests/python/unittest/test_gluon.py]: According to python docs, it seems to be deleted by the user. [The user of mkdtemp() is responsible for deleting the temporary directory and its contents when done with it.](https://docs.python.org/2/library/tempfile.html#tempfile.mkdtemp)

Comment (sandeep-krishnamurthy) [reply tests/python/unittest/test_gluon.py]: Right. Thanks. I meant, temp gets automatically cleaned up after all the tests (test session).

Comment (sandeep-krishnamurthy) [reply tests/python/unittest/test_gluon.py]: @marcoabreu - Can you please confirm if my understanding is correct? If not, I will add code to delete the temp directory created in the tests. Also, I see similar behavior in all other tests, where it creates temp dir, but, assumes, it will be cleaned up the system.

**Comment by sandeep-krishnamurthy** [comment]
@zhreshold - Can you help look at this once, and merge if looks good?

File: python/mxnet/gluon/block.py [hunk 7]
```diff
@@ -1086,3 +1094,54 @@ def _clear_cached_op(self):
...
+    aux_types = None
+
+    # Get Input symbol details. This will be used to infer types of
+    # other parameters.
+    input_sym_name = in_params.name
+    input_sym_arg_type = in_params.infer_type()[0]
+
+    # Try to infer types of other parameters.
+    if input_sym_arg_type and len(input_sym_arg_type) > 0:
+        params = {input_sym_name:input_sym_arg_type[0]}
```
Comment (zhreshold) [inline python/mxnet/gluon/block.py]: this does not handle Grouped Symbol because you are slicing [0] from the symbol.

File: python/mxnet/gluon/block.py [hunk 8]
```diff
@@ -1053,13 +1054,20 @@ def __init__(self, outputs, inputs, params=None):
...
-                self.params.get(i, allow_deferred_init=True)
+        # Infer type of parameters. Without this, every parameter will be created with
+        # default type i.e., fp32
+        arg_params = out.list_arguments()
+        aux_params = out.list_auxiliary_states()
 
-        for i in out.list_auxiliary_states():
-            if i not in input_names:
-                self.params.get(i, grad_req='null', allow_deferred_init=True)
+        arg_types, aux_types = _infer_param_types(inputs[0], out, arg_params, aux_params)
```
Comment (zhreshold) [inline python/mxnet/gluon/block.py]: this does not handle Grouped Symbol because you are only slicing [0] from the symbol.

Comment (zhreshold) [reply python/mxnet/gluon/block.py]: also, I think the type inference should occur in `cast` as well, otherwise it's buggy when user is trying to cast dtype of a cascaded network with symbolBlock inside.

File: tests/python/unittest/test_gluon.py [hunk 9]
```diff
@@ -336,6 +339,41 @@ def hybrid_forward(self, F, x):
...
+    net_fp64 = mx.gluon.SymbolBlock(sm, inputs)
+    net_fp64.collect_params().load(tmpfile + '-0000.params', ctx=ctx)
+    # 3. Get a conv layer's weight parameter name. Conv layer's weight param is
+    # expected to be of dtype casted, fp64.
+    for param_name in net_fp64.params.keys():
+        if 'conv' in param_name and 'weight' in param_name:
+            break
+    assert np.dtype(net_fp64.params[param_name].dtype) == np.dtype(np.float64)
+
+    # Cast the symbol block to FP32 and try to forward a FP32 data.
```
Comment (sandeep-krishnamurthy) [inline tests/python/unittest/test_gluon.py]: @zhreshold - Added a test here to verify the SymbolBlock.cast functionality.

**Comment by sandeep-krishnamurthy** [comment]
@zhreshold - Can you please take a look at this? Thanks.

File: python/mxnet/gluon/block.py [hunk 10]
```diff
@@ -1084,5 +1092,71 @@ def _clear_cached_op(self):
...
+        Defaults to 'float32', if unable to infer type.
+    """
+    arg_types = None
+    aux_types = None
+
+    # Get Input symbol details. This will be used to infer types of
+    # other parameters.
+    input_sym_names = [in_param.name for in_param in in_params]
+
+    # Try to infer input types. If not successful, we will set default dtype.
```
Comment (zhreshold) [inline python/mxnet/gluon/block.py]: could you elaborate when infer type is not successful, and should we raise if that happens?

Comment (sandeep-krishnamurthy) [reply python/mxnet/gluon/block.py]: In one of the unit test(pasting below) infer shape will be false and it need not be an issue, users are just assuming default type.

```python
    model = nn.HybridSequential()
    model.add(nn.Dense(128, activation='tanh'))
    model.add(nn.Dropout(0.5))
    model.add(nn.Dense(64, activation='tanh'),
              nn.Dense(32, in_units=64))
    model.add(nn.Activation('relu'))

    model.initialize()

    inputs = mx.sym.var('data')
    outputs = model(inputs).get_internals()

    smodel = gluon.SymbolBlock(outputs, inputs, params=model.collect_params())
```

Comment (zhreshold) [reply python/mxnet/gluon/block.py]: I don't get why this would fail to infer_type?

Comment (sandeep-krishnamurthy) [reply python/mxnet/gluon/block.py]: Because, input symbol is not created with dtype.
```python
>>> data = mx.sym.var('data')
>>> data.infer_type()
(None, None, None)
```
```python
>>> data = mx.sym.var('data', dtype='float32')
>>> data.infer_type()
([<class 'numpy.float32'>], [<class 'numpy.float32'>], [])
```
